src/sage/groups/additive_abelian/additive_abelian_wrapper.py

```diff
@@ -566,3 +616,188 @@ def _rec(j, k, c):
         return x
 
     return _rec(0, max(vals), b)
+
+
+def _expand_basis_pgroup(p, alphas, vals, beta, h, rel):
+    r"""
+    Given a basis of a `p`-subgroup of a finite abelian group
+    and an element lying outside the subgroup, extend the basis
+    to the subgroup spanned jointly by the original subgroup and
+    the new element.
+    Used as a subroutine in :func:`basis_from_generators`.
+
+    This function modifies ``alphas`` and ``vals`` in place.
+
+    ALGORITHM: [Suth2007]_, Algorithm 9.2
+
+    EXAMPLES::
+
+        sage: from sage.groups.additive_abelian.additive_abelian_wrapper import _expand_basis_pgroup
+        sage: A = AdditiveAbelianGroup([9,3])
+        sage: alphas = [A((5,2))]
+        sage: beta = A((1,0))
+        sage: vals = [2]
+        sage: rel = next([ZZ(r),ZZ(s)] for s in range(9) for r in range(9) if s > 1 and not r*alphas[0] + s*beta)
+        sage: _expand_basis_pgroup(3, alphas, vals, beta, 2, rel)
+        sage: alphas
+        [(5, 2), (6, 2)]
+        sage: vals
+        [2, 1]
+        sage: len({i*alphas[0] + j*alphas[1] for i in range(3^2) for j in range(3^1)})
+        27
+    """
+    k = len(rel)
+    assert isinstance(alphas, list) and isinstance(vals, list)
+    assert len(alphas) == len(vals) == k-1
+    assert all(r >= 0 for r in rel)
+#    assert not sum(r*a for r,a in zip(rel, alphas+[beta]))
+#    assert all(a.order() == p**v for a,v in zip(alphas,vals))
+
+    # step 1
+    for i in range(k-1):
+        if not rel[i]:
+            continue
+        q = rel[i].p_primary_part(p)
+        alphas[i] *= rel[i]//q
+        rel[i] = q
+    min_r = min(filter(bool, rel))
+    val_rlast = rel[-1].valuation(p)
+    assert rel[-1] == p**val_rlast
```

Comment:
```suggestion
    assert rel[-1] == p ** val_rlast
```
PEP8 (I also think it is dumb, but we should spare Frederic from having to make such a change, or the linter complaining).